cmd/frontend/internal/frontend/service.go

```diff
@@ -258,25 +258,30 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
 		return
 	}
 
-	logForced := func() bool {
-		fes.monitorMu.Lock()
-		defer fes.monitorMu.Unlock()
-		forced := fes.logNextMonitorStatus
-		fes.logNextMonitorStatus = false
-		return forced
-	}
-
-	//linkCh := make(chan string, 1)
 	go func() {
-		var once sync.Once
-		//defer close(linkCh)
 		// status of external connectivity; requires 1 GW per IP family if configured to be reachable
-		init, noConnectivity := true, true
+		hasConnectivity := false
 		connectivityMap := map[string]bool{}
 		delay := 3 * time.Second // when started grant more time to write the whole config (minimize intial link flapping)
 		_ = fes.WaitStart(ctx)
 		var refreshCancel context.CancelFunc
+		denounce := true // Always denounce on container start
 		for {
+
+			if denounce {
+				denounce = false
+				hasConnectivity = false
+				if refreshCancel != nil {
+					refreshCancel()
+					refreshCancel = nil
+				}
+				if err := denounceFrontend(ctx, fes.targetRegistryClient); err != nil {
```

Comment:
> or some simple but lame init delay could be used.

I go for this option as a start. Same as before, a 3s delay

Reply:
Seems that in the baseline it wasn't the delay rather the grpc option WaitForReady that alowed the initial (or any) denounce to succeed. It blocks RPCs until the connection is ready.

Even with the 3 seconds delay, denounce fails on xcluster for me.

I think the easiest way forward would be if denounce did not have hard coded timeout, instead it would be the callers responsibility to pass a feasible context. Or maybe just increase the hard-coded value in denounce to 30 seconds, and in case of shutdown pass a context with lower timeout.

Reply:
I think it may be best just to remove the timeout in announceFrontend/denounceFrontend entirely and make a comment. Not adding a ctx with timeout at all in the calls. I.e. same as it was before. Timeouts would be hard to decide, especially if they are dynamic. The passed ctx can still be canceled from main.

Or do you think there is a case where the calls can hang for an unreasonable time (whatever that is)?

Reply:
> and in case of shutdown pass a context with lower timeout.

I missed this. Of course, a timeout is needed on shutdown

Reply:
You're right. Expect for the shutdown, I don't see much need for timeout.